Fix/fix button state

valeryV99 opened the pull request:
Added tests for form components
improved useSwapRates
https://www.notion.so/254362757c98814193bdf2683a2f8535?v=254362757c98811fb5a9000c47b44575&p=26a362757c988020be2ccffed6192858&pm=s

valeryV99 committed:
feat: enhance SwapForm with usability and testing improvements

- Disabled swap inputs and submit button when rates are fetching.
- Ensured consistent state handling by moving `isFetching` reset to `finally` block in `useSwapRates`.
- Increased test coverage:
  - Added unit tests for `useSwapRates`.
  - Introduced integration tests for `SwapForm` to validate swap interactions and edge cases.
- Expanded Vitest coverage configuration to include `SwapForm` and `useSwapRates`.
- Minor style updates, e.g., `minWidth` for primary button in `SwapForm`.

valeryV99 committed:
Revert "feat: enhance SwapForm with usability and testing improvements"

This reverts commit eabbcb608fd9eddebd0d4be5b5439246e2d513a2.

valeryV99 committed:
Merge remote-tracking branch 'origin/staging' into fix/fix-button-state

valeryV99 committed:
feat: improve SwapForm usability and enhance testing

- Disable swap inputs and submit button during rate fetching for better UX.
- Ensure proper `isFetching` state management by moving reset logic to `finally` block.
- Add comprehensive unit tests for `useSwapRates` including error handling and reverse calculations.
- Introduce integration tests for `SwapForm` covering swap interactions and edge cases.
- Expand Vitest coverage configuration to include `SwapForm` and `useSwapRates`.
- Make minor UI adjustments, e.g., set `minWidth` for primary button.

opcode0x commented on apps/web/src/features/swap/api/useSwapRates.test.tsx:
```diff
@@ -0,0 +1,103 @@
+describe('useSwapRates (renderHook)', () => {
+  beforeEach(() => {
+    readContractMock.mockReset();
+    cleanup();
```
You can void manual clean up by adding test.globals: true in vitest configuration

valeryV99 replied:
Thanks

opcode0x commented on apps/web/src/widgets/swap-form/ui/SwapForm.test.tsx:
```diff
@@ -0,0 +1,370 @@
+vi.mock('@/shared', async () => {
+  const { useFormContext } = await import('react-hook-form');
+  return {
+    Button: ({ children, type = 'button', ...props }: any) => (
```
Why do we need to mock the button ?

valeryV99 replied:
In the mock I left minimal semantics button to focus on the logic of the form, not on the implementation of the button.
Reworked.

opcode0x commented on apps/web/src/widgets/swap-form/ui/SwapForm.test.tsx:
```diff
@@ -0,0 +1,370 @@
+        {children}
+      </button>
+    ),
+    parseNumberInput: (v: string) => v,
```
Why do we need to mock this ?

valeryV99 replied:
`+`

opcode0x commented on apps/web/src/widgets/swap-form/ui/SwapForm.test.tsx:
```diff
@@ -0,0 +1,370 @@
+      </button>
+    ),
+    parseNumberInput: (v: string) => v,
+    SwapInput: ({ amountName, onChange, ...rest }: any) => {
```
Let's avoid mocking swap input. We can get the element by its role or aria label

valeryV99 replied:
`+`

valeryV99 committed:
feat: update Vitest configs and refactor SwapForm tests

- Enabled `globals` in Vitest configuration for simplified test setup.
- Switched to using `fireEvent.input` instead of `fireEvent.change` for consistent event simulation.
- Removed redundant cleanup calls from tests to streamline test logic.
- Refactored mocked modules to improve readability and maintainability.
- Added minimal `TokenButton` mock for testing `SwapInput` rendering.
- Enhanced test readability with cleaner mock and logic updates.

valeryV99 committed:
Merge branch 'staging' into fix/fix-button-state

valeryV99 committed:
Merge remote-tracking branch 'origin/staging' into fix/fix-button-state

# Conflicts:
#	apps/web/vitest.config.mts
